### ISSUE TITLE
Serializing blank nodes not working?

### ISSUE THREAD

**Issue by pdenno** [ontodev/edn-ld#1]
It appears that blank nodes are not serialized (at all). 

(use 'edn-ld.core 'edn-ld.common 'edn-ld.jena)
(def context+ (merge default-context {:mfi "http://modelmeth.nist.gov/mfi/", nil :mfi}))
(def prefixes (assoc (get-prefixes context+) :rdf rdf :xsd xsd))
(println (edn-ld.jena/write-triple-string 
             prefixes
             (expand-all context+ '(["_:foo" :rdf:type :MathRelation]))))

@prefix mfi:   http://modelmeth.nist.gov/mfi/ .
@prefix rdf:   http://www.w3.org/1999/02/22-rdf-syntax-ns# .
@prefix owl:   http://www.w3.org/2002/07/owl# .
@prefix xsd:   http://www.w3.org/2001/XMLSchema# .
@prefix rdfs:  http://www.w3.org/2000/01/rdf-schema# .

[ a       mfi:MathRelation ] .

**Comment by jamesaoverton** [comment]
Thanks for the report. I can confirm the behaviour.

I poked around with this. The behaviour is not what I expected, but I don't think it's a bug so much as differing expectations about how blank nodes should be represented. In general, I recommend against using blank nodes, for this reason and several other reasons.

I believe that I'm using Jena correctly to create the blank nodes. In this case, I would like to see the triple:

```
_:foo a mfi:MathRelation .
```

But using the square brackets is valid Turtle syntax for representing a subject that is a blank node. Jena is not keeping the blank node identifiers that I'm giving it, but that's probably their expected behaviour.

I discovered other bugs with jena.clj, and I'm going to more research over the next few weeks, so I'll leave this issue open for now. But my recommendation is that, if this is not acceptable behaviour for your use case, that you either avoid using blank nodes, or write some custom Turtle serialization code (which should not be hard).

**Comment by pdenno** [comment]
Perhaps my example was not quite sufficient. However, given the following: 

```
(defn new-blank-node   []  (str "_:" (gensym "mfi")))

(def expanded-math
  (let [subj-rel1 (new-blank-node)
        subj-var1 (new-blank-node)
        subj-var2  (new-blank-node)]
    (map #(expand-all context+ %) 
        `([~subj-rel1 :rdf:type :MathRelation]
           [~subj-rel1 :hasLHS ~subj-var1]
           [~subj-var1 :rdf:type :MathVariable]
         ; [~subj-var1 :ofRelation ~subj-rel1]   ; I added this one.
           [~subj-var1 :hasVarName {:value "Y"}]
           [~subj-rel1 :hasRelationOp :MathRelationEquality]
           [~subj-rel1 :hasRHS ~subj-var2]
           [~subj-var2 :rdf:type :MathVariable]
           [~subj-var2 :hasVarName {:value "X"}]))))
```

If you remove the ; from the line commented "I added this one" then a blank node used as a subject is now also used as an object. No triples are serialized in that case:

```
mfi.core> (println (edn-ld.jena/write-triple-string prefixes expanded-math))
@prefix mfi:   <http://modelmeth.nist.gov/mfi/> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .


nil
mfi.core> 
```

This page on RIOT: https://jena.apache.org/documentation/io/rdf-output.html describes some aspects of using blank nodes, including this: "RDF lists are printed as (...) and [...] is used for blank nodes where possible." reinforcing your point made earlier. But I don't see any limitation described there that would alert one to the behavior shown above. 

I believe that blank nodes are an important element of the RDF language. They are the best (only?) means to express an existential that you do not wish to denote.

**Comment by jamesaoverton** [comment]
Ok, thanks for the extended example. I'm travelling today, without proper Internet access. I'll look at this more closely as soon as I can.

**Comment by pdenno** [comment]
The original version of the message that I saved on the github 
interpreted a backquote as markdown and hid some code. I fixed that.

**Comment by jamesaoverton** [comment]
Thanks again for the bug report. It's proving to be a complicated one.

I can confirm the behaviour of your second example. However, writing to other formats works for me:

```
(println (edn-ld.jena/write-triple-string "RDFXML" prefixes expanded-math))
(println (edn-ld.jena/write-triple-string "JSON-LD" prefixes expanded-math))
```

Can you confirm that RDFXML and JSON-LD formats work for you? If so, I think I'm using Jena correctly, and this is a bug on their end with blank node nesting for Turtle. Maybe there's an existing ticket, I haven't had time to check. If not, it would be good to isolate a minimal case and file a report with Jena.

I tried to reproduce the bug with pure Jena using their `riot` command line tool. But `riot` works fine reading/writing this graph from/to TTL, because it's explicit about the blank nodes (no "[]" blocks). Now I see that `riot` only uses the streaming RDF formats, which do not include RDFXML and JSON-LD. EDN-LD could write using the streaming RDF APIs, but I want to support the full range of formats. Maybe a hybrid approach is possible.

**Comment by pdenno** [comment]
I can confirm that the both RDFXML and JSON-LD work. The way the RDFXML is serialized -- by flipping  things around so that the mfi:MathRelation element is a child of the mfi:MathVariable that references it -- makes unnecessary a blank node for the subject of what I was seeing as the "root node," the MathRelation. That got me thinking that maybe the turtle serialization code is having trouble with my example because **every** node is a blank node. That looks to be the case! If I nail down the MathRelation with an IRI, I get valid turtle. I can even force use of _:foo syntax in subjects and object. Here is an example:

```
(def expanded-math
  (let [subj-rel1 :realURI ; (new-blank-node)
        subj-var1 (new-blank-node)
        subj-var2  (new-blank-node)]
    (map #(expand-all context+ %) 
         `([~subj-rel1 :rdf:type :MathRelation]
           [~subj-rel1 :hasLHS ~subj-var1]
           [~subj-var1 :rdf:type :MathVariable]
           [~subj-var1 :ofRelation ~subj-rel1]  
           [~subj-var1 :hasBN1 "_:foobar1"]     ; I added this one 
           [~subj-var1 :hasBN2 "_:foobar2"]     ; I added this one 
           [~subj-var1 :hasVarName {:value "Y"}]
           [~subj-rel1 :hasRelationOp :MathRelationEquality]
           [~subj-rel1 :hasBN2  "_:foobar2"]    ; I added this one
           [~subj-rel1 :hasRHS ~subj-var2]
           [~subj-var2 :rdf:type :MathVariable]
           [~subj-var2 :otherside ~subj-var1] ; I added this one
           [~subj-var2 :hasVarName {:value "X"}]))))
```

mfi.core> (println (edn-ld.jena/write-triple-string prefixes expanded-math))

```
@prefix mfi:   <http://modelmeth.nist.gov/mfi/> .
@prefix rdf:   <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix owl:   <http://www.w3.org/2002/07/owl#> .
@prefix xsd:   <http://www.w3.org/2001/XMLSchema#> .
@prefix rdfs:  <http://www.w3.org/2000/01/rdf-schema#> .

_:b0    a               mfi:MathVariable ;
        mfi:hasBN1      []  ;
        mfi:hasBN2      _:b1 ;
        mfi:hasVarName  "Y"^^xsd:string ;
        mfi:ofRelation  mfi:realURI .

mfi:realURI  a             mfi:MathRelation ;
        mfi:hasBN2         _:b1 ;
        mfi:hasLHS         _:b0 ;
        mfi:hasRHS         [ a               mfi:MathVariable ;
                             mfi:hasVarName  "X"^^xsd:string ;
                             mfi:otherside   _:b0
                           ] ;
        mfi:hasRelationOp  mfi:MathRelationEquality .
```

So, I guess I'd conclude that though my previous example and your experiments with RDFXML and JSON-LD suggest a bug with Jena turtle serialization, it isn't a very big bug. At least not big if it only takes one IRI to get things working. I'll be playing with the code over the next few weeks; maybe that will shed more light on it. Thanks for all the help on this.

**Comment by jamesaoverton** [comment]
This Jena tracker item looks right: https://issues.apache.org/jira/browse/JENA-908. It's marked fixed in 2.13.1, which has not yet been released. I'll leave our tracker item open until I can update the dependency.

**Comment by pdenno** [comment]
That looks like it! Good to hear.
